refactor(sign): move legacy signs to extmarks

luukvbaal opened the pull request:
**Problem**:  The legacy signlist data structures and associated functions are
          redundant since the introduction of extmark signs.

**Solution**: Store signs defined through the legacy commands in a hashmap, placed
          signs in the extmark tree. Replace signlist associated functions.

---

Usage of the legacy sign commands should yield no change in behavior with the
exception of:
  - "orphaned signs" are now always removed when the line it is placed on is
    deleted. This used to depend on the value of 'signcolumn'.
  - It is no longer possible to place multiple signs with the same identifier
    in a single group on multiple lines. This will now move the sign instead.

Moreover, both signs placed through the legacy sign commands and through
|nvim_buf_set_extmark()|:
  - Will show up in both |sign-place| and |nvim_buf_get_extmarks()|.
  - Are displayed by increasing sign identifier, left to right.
    Extmark signs used to be ordered decreasingly as opposed to legacy signs.

lewis6991 commented:
Good work. This looks awesome!

luukvbaal commented:
I believe all issues related to sign ordering are now resolved. The few display tests of which I changed the expected result were modified to match the extmark sign order to legacy sign order when it is not determined by `priority` but by `sign id` (as discussed and suggested by @bfredl on matrix).

The remaining failing tests require a solution for extmark signs to match legacy behavior where deleted lines have their associated signs removed from the buffer.

luukvbaal commented:
Extmark signs will currently show up in `sign place group=*` and signs placed through the legacy commands will show up in `nvim_buf_get_extmarks()`. IMO it's not really a problem, more so a useful (unintended) consequence of this refactor. If anyone disagrees let me know and we can keep them segregated.

bfredl commented on src/nvim/extmark.c:
```diff
@@ -156,12 +159,16 @@ void extmark_set(buf_T *buf, uint32_t ns_id, uint32_t *idp, int row, colnr_T col
       buf->b_virt_line_blocks++;
     }
     if (decor_has_sign(decor)) {
+      if (revised) {
+        decor->sign_mod_id = sign_add_id;
+      }
+      decor->sign_add_id = sign_add_id++;
```
How much stuff would break if this was reduced to a single id (which would be incremented regardless if the sign is newly added or modified)?

luukvbaal replied:
Arguably not that much. When signs are first placed in the signlist currently they are sorted by `lnum` -> `priority` -> `sign_id`. Then when modified signs take precedence over other signs with the same `sign_id`.
https://github.com/neovim/neovim/blob/5e5f5174e3faa862a9bc353aa7da41487911140b/src/nvim/sign.c#L309-L312

Tested e.g. here:
https://github.com/neovim/neovim/blob/5e5f5174e3faa862a9bc353aa7da41487911140b/test/old/testdir/test_signs.vim#L1512-L1515

bfredl replied:
IMO we should change this and just keep track of the last modified time. This is the kind of bug-for-bug compatibility which we don't aim at supporting.

luukvbaal replied:
To be clear, we then want signs to be sorted by `lnum`->`priority`->`sign_id`->`last_modifed`?

bfredl replied:
>  then want signs to be sorted by lnum->priority->sign_id->last_modifed?

Sounds reasonable

luukvbaal commented:
> Add a new (public?) extmark property to have the mark removed when the line it is placed on is deleted to match the legacy sign behavior. Should fix the last remaining failing tests?

I'm kind of stuck on this. First of all, do we want to keep sign deletion conditional on the value of [`'signcolumn'`](https://github.com/neovim/neovim/blob/94127cb5df0a513e66777d18a2c7fa6219404280/runtime/doc/options.txt#L5620-L5624)?

Does this need to be a publicly configurable property for all extmarks? Or can it simply be done implicitly for marks containing sign properties? And would looping over all marks in the deleted range and simply deleting when the flag is set be fine? I think if done properly it would be handled somewhere in `marktree_splice()` but I'm not sure I would manage that.

Can this note about undoing sign deletion be removed, or does this need to be maintained along with the behavior somehow? Signs being extmarks now move back upon undo'ing.
https://github.com/neovim/neovim/blob/94127cb5df0a513e66777d18a2c7fa6219404280/runtime/doc/sign.txt#L80-L82

luukvbaal commented:
> Ideally all marks. But if it makes it easier, you can add the logic internally and defer adding this to the public extmark API. I was thinking this could be added as `lifetime = 'buffer' | 'redraw'` and deprecate `ephemeral`.

Alright, I think I'll defer it for now then. Just need to keep in mind that whatever code I add now might become redundant whenever it is handled properly for all marks.

> > And would looping over all marks in the deleted range and simply deleting when the flag is set/is a sign be fine? I think if done properly it would be handled somewhere in marktree_splice() but I'm not sure I would manage that.
> 
> Yes I believe so. Can't you just mimic the code for what we do for signs?

Yeah this is no problem, and is what I had already implemented locally. I just believe its not idiomatic with the rest of the extmark adjusting code.

> > Can this note about undoing sign deletion be removed, or does this need to be maintained along with the behavior somehow? Signs (being extmarks) now move back upon undo'ing.
> 
> Isn't that note superseded by the fact that signs are deleted when lines are removed?

Correct. If we get rid of the conditional removal of signs and instead always remove them, this note is redundant all together.

luukvbaal commented:
> > > And would looping over all marks in the deleted range and simply deleting when the flag is set/is a sign be fine? I think if done properly it would be handled somewhere in marktree_splice() but I'm not sure I would manage that.
> > 
> > 
> > Yes I believe so. Can't you just mimic the code for what we do for signs?
> 
> Yeah this is no problem, and is what I had already implemented locally. I just believe its not idiomatic with the rest of the extmark adjusting code.

Hmm it does result in problems when joining lines. In `do_join()`, [`del_lines()`](https://github.com/neovim/neovim/blob/94127cb5df0a513e66777d18a2c7fa6219404280/src/nvim/ops.c#L4172C7-L4172C7) is called after extmarks have already been adjusted by [`extmark_splice()`](https://github.com/neovim/neovim/blob/94127cb5df0a513e66777d18a2c7fa6219404280/src/nvim/ops.c#L4172C7-L4172C7). This results in marks being deleted in lines succeeding the joined lines with my current implementation.

I get the feeling deleting flagged marks should be handled while iterating in the splice routine. But TBH I haven't wrapped my head around what really happens in `extmark_splice()`. So IDK if I can manage that, or if my intuition is even correct.

luukvbaal commented:
OK found something that appears to work properly, also for joining lines. I guess I can take a stab at making a separate PR for the API change. @lewis6991 I didn't quite get how you wanted `lifetime` to work, `buffer|redraw` seem to me to be the current default mark lifetime and the lifetime for `ephemeral` marks. Were you suggesting we add those in addition to a third `lifetime` e.g. `line`, or am I missing something?

```
- lifetime : control the lifetime of the mark.
  - "buffer": the mark will be stored in the buffer and is only
              removed when explicitly done so with
              |nvim_buf_del_extmark()|.
  - "line":   the mark will be stored in the buffer and is removed
              when the line it is placed on is deleted.
              explicitly done so with |nvim_buf_del_extmark()|.
  - "redraw": the mark will only be used for the current redraw
              cycle, and not be permanently stored in the buffer.
              To be used with |nvim_set_decoration_provider()|.
```

luukvbaal commented:
I have included the invalidation flags from #25767 here (and made it to work with undo), and set them implicitly for sign extmarks. Marks that were placed on deleted lines are not drawn, and are restored when the deletion is undone.

If that behavior is desireable as an API flag I'd be glad to revive #25767 but after the discussion there I'm not so sure anymore.

luukvbaal commented:
Might be room for some more cleanup and this will surely conflict with #25826 (IDK if that will merge before this), but I don't think the general shape of this PR will change anymore so accepting reviews.

bfredl commented:
> IDK if that will merge before this

I haven't strictly decided on that yet. But if this is is close to ready maybe we should get this in first, and I can do the rebase fixes (shouldn't be a big deal, just shuffle things around).

luukvbaal commented on src/nvim/buffer.c:
```diff
@@ -4149,42 +4092,32 @@ void buf_signcols_add_check(buf_T *buf, sign_entry_T *added)
-  // maximum.
-  if (maximum > buf->b_signcols.max) {
+  // (valid) maximum.
+  if (buf->b_signcols.max && max > buf->b_signcols.max) {
```
This was flip-flopping `b_signcols.valid` in a buffer without any signs. Causing unnecessarily looping over all marks in `decor_signcols()` each time any extmark was added. (Could try adding a regression test for this, maybe through FFI?)

bfredl replied:
good catch. Doesn't need a specific test.

luukvbaal commented on src/nvim/globals.h:
```diff
@@ -476,10 +476,6 @@ EXTERN buf_T *curbuf INIT( = NULL);    // currently active buffer
   for ((wip) = (buf)->b_wininfo; (wip) != NULL; (wip) = (wip)->wi_next)   // NOLINT
 
-// Iterate through all the signs placed in a buffer
-#define FOR_ALL_SIGNS_IN_BUF(buf, sign) \
```
I wonder if an elaborate `FOR_ALL_MARKS` macro or function with function pointer argument would be useful for reducing boilerplate. Or would it only obfuscate unnecessarily?

bfredl replied:
Shouldn't be a need, direct control of the iterator is better and is not much boiler plate in the simple case.

luukvbaal commented on src/nvim/sign.c:
```diff
@@ -612,206 +193,139 @@ static linenr_T buf_delsign(buf_T *buf, linenr_T atlnum, int id, char *group)
+        kv_push(signs, mark);
+        marktree_itr_next(buf->b_marktree, itr);
+      } else {
+        extmark_del(buf, itr, mark, 0, 0);
```
Is this `extmark_del()` refactor alright and if so is there any reason why it shouldn't be used in `extmark_clear()`? I noticed that adds all marks/decors to a vector instead of deleting during iteration.

bfredl commented on runtime/doc/sign.txt:
```diff
@@ -77,9 +75,9 @@ When two signs with the same priority are present, and one has an icon or text
-the delete is undone the sign does not move back.
+When the line on which the sign is placed is deleted, the sign is removed along
+with it. This is in contrast to the default Vim behavior of keeping and grouping
+them. Undoing the line deletion also restores the removed signs.
```
It is fine to label the commit as `refactor()` but would still be good with a brief news.txt item acknowledging that the behavior of legacy signs has changed.

lewis6991 commented on runtime/doc/news.txt:
```diff
@@ -80,6 +80,14 @@ The following changes may require adaptations in user config or plugins.
 
+• Signs placed through the legacy |sign-commands| are now stored and displayed
+  as |extmarks| internally. Along with the following changes:
+  • Lines deleted and restored with |undo| have their signs removed and restored.
```
I think this can be considered a breaking change and will require me to make changes to several of my plugins.

bfredl replied:
yea, if we add the "ignore undo" mode (I'll put up a PR), it would make sense to use that for legacy signs.

luukvbaal replied:
Ok so what do we want to happen in this PR upon line deletion?

1. Both legacy and extmark signs will be deleted from the buffer (regardless of 'signcolumn', "breaking" for extmark signs).
2. We make a superficial distinction between extmark and legacy signs and have extmark signs adjusted and legacy signs deleted, just to stay backwards compatible.
3. Both legacy and extmark signs will be adjusted instead of deleted ("breaking" for legacy signs).

bfredl replied:
The distinction will rather be that extmark signs _explicitly_ can choose either mode, just like any other extmark.

"legacy" behavior becomes a strict subset of the configurable extmark behavior in either case.

luukvbaal replied:
Yeah fair, I guess we're set then once I rebase this on top of #25882. Legacy signs will have both `invalidate` and `no_undo_restore` set and will be deleted, only breaking the current `'signcolumn' == yes/auto:1` behavior.

luukvbaal commented on src/nvim/decoration.c:
```diff
@@ -105,12 +107,13 @@ void decor_add(buf_T *buf, int row, int row2, Decoration *decor, bool hl_id)
     }
     if (decor->sign_text) {
       buf->b_signs_with_text++;
       // TODO(lewis6991): smarter invalidation
```
Can this be removed now that the sentinel line is being checked for extmark signs in `buf_signcols_add_check()`, or did you plan something else @lewis6991?

lewis6991 replied:
Maybe? I can't really remember without deep diving. If you think it's optimal (...enough) then feel free to remove.

luukvbaal committed:
refactor(sign): move legacy signs to extmarks

Problem:  The legacy signlist data structures and associated functions are
          redundant since the introduction of extmark signs.
Solution: Store signs defined through the legacy commands in a hashmap, placed
          signs in the extmark tree. Replace signlist associated functions.

Usage of the legacy sign commands should yield no change in behavior with the
exception of:
  - "orphaned signs" are now always removed when the line it is placed on is
    deleted. This used to depend on the value of 'signcolumn'.
  - It is no longer possible to place multiple signs with the same identifier
    in a single group on multiple lines. This will now move the sign instead.

Moreover, both signs placed through the legacy sign commands and through
|nvim_buf_set_extmark()|:
  - Will show up in both |sign-place| and |nvim_buf_get_extmarks()|.
  - Are displayed by increasing sign identifier, left to right.
    Extmark signs used to be ordered decreasingly as opposed to legacy signs.

justinmk commented:
> will show up in both |sign-place| and |nvim_buf_get_extmarks()|.

worth adding a hint to the nvim_buf_get_extmarks docstring:

```diff
diff --git a/src/nvim/api/extmark.c b/src/nvim/api/extmark.c
index 84b89a7428cb..36b273eda2ae 100644
--- a/src/nvim/api/extmark.c
+++ b/src/nvim/api/extmark.c
@@ -314,7 +314,7 @@ ArrayOf(Integer) nvim_buf_get_extmark_by_id(Buffer buffer, Integer ns_id,
   return extmark_to_array(&extmark, false, details, hl_name);
 }
 
-/// Gets |extmarks| in "traversal order" from a |charwise| region defined by
+/// Gets |extmarks| (including |signs|) in "traversal order" from a |charwise| region defined by
 /// buffer positions (inclusive, 0-indexed |api-indexing|).
 ///
 /// Region can be given as (row,col) tuples, or valid extmark ids (whose
```

lewis6991 commented on test/old/testdir/test_signs.vim:
```diff
@@ -1670,34 +1674,33 @@ func Test_sign_lnum_adjust()
+  " " Delete the line with the sign
+  " call deletebufline('', 4)
+  " let l = sign_getplaced(bufnr(''))
+  " call assert_equal(4, l[0].signs[0].lnum)
```
Can this test simply be updated (or removed) instead of commenting it?

zeertzjq replied:
No.

lewis6991 replied:
If I understand correctly. The commented out test here isn't even vim behaviour which moves the signs instead.

luukvbaal replied:
This test tests vim behavior, which we matched by setting `'signcolumn'` to `"yes:1". Since we decided we no longer want conditional sign removal, we can no longer use this test. Adapting it seems unnecessary, as signs now just match extmark behavior which is well tested. Apparently commenting it out is preferred to removing?

luukvbaal replied:
The other test could be adapted I guess.

zeertzjq replied:
I think yes. Commenting out two lines and adding two lines is a smaller change than commenting out all assertions. Note that there cannot be spaces between `"\` though.

luukvbaal replied:
Not sure what you mean, what I had in mind was asserting that replacing those signs doesn't change the order:
```diff
diff --git a/test/old/testdir/test_signs.vim b/test/old/testdir/test_signs.vim
index 507efcefe..1beace7d3 100644
--- a/test/old/testdir/test_signs.vim
+++ b/test/old/testdir/test_signs.vim
@@ -1501,54 +1501,33 @@ func Test_sign_priority()
   call sign_place(3, '', 'sign3', 'Xsign',
               \ {'lnum' : 4, 'priority' : 20})
   let s = sign_getplaced('Xsign', {'group' : '*'})
-  call assert_equal([
+  let se = [
               \ {'id' : 3, 'name' : 'sign3', 'lnum' : 4, 'group' : '',
               \ 'priority' : 20},
               \ {'id' : 2, 'name' : 'sign2', 'lnum' : 4, 'group' : '',
               \ 'priority' : 20},
               \ {'id' : 1, 'name' : 'sign1', 'lnum' : 4, 'group' : '',
-              \ 'priority' : 20}],
-              \ s[0].signs)
+              \ 'priority' : 20}]
+  call assert_equal(es, s[0].signs)
 
   " Nvim: signs are always sorted lnum->priority->sign_id->last_modified
   " Last modified does not take precedence over sign_id here.
 
   " Place the last sign again with the same priority
-  " call sign_place(1, '', 'sign1', 'Xsign',
-  "             \ {'lnum' : 4, 'priority' : 20})
-  " let s = sign_getplaced('Xsign', {'group' : '*'})
-  " call assert_equal([
-  "             \ {'id' : 1, 'name' : 'sign1', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20},
-  "             \ {'id' : 3, 'name' : 'sign3', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20},
-  "             \ {'id' : 2, 'name' : 'sign2', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20}],
-  "             \ s[0].signs)
-  " " Place the first sign again with the same priority
-  " call sign_place(1, '', 'sign1', 'Xsign',
-  "             \ {'lnum' : 4, 'priority' : 20})
-  " let s = sign_getplaced('Xsign', {'group' : '*'})
-  " call assert_equal([
-  "             \ {'id' : 1, 'name' : 'sign1', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20},
-  "             \ {'id' : 3, 'name' : 'sign3', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20},
-  "             \ {'id' : 2, 'name' : 'sign2', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20}],
-  "             \ s[0].signs)
-  " " Place the middle sign again with the same priority
-  " call sign_place(3, '', 'sign3', 'Xsign',
-  "             \ {'lnum' : 4, 'priority' : 20})
-  " let s = sign_getplaced('Xsign', {'group' : '*'})
-  " call assert_equal([
-  "             \ {'id' : 3, 'name' : 'sign3', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20},
-  "             \ {'id' : 1, 'name' : 'sign1', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20},
-  "             \ {'id' : 2, 'name' : 'sign2', 'lnum' : 4, 'group' : '',
-  "             \ 'priority' : 20}],
-  "             \ s[0].signs)
+  call sign_place(1, '', 'sign1', 'Xsign',
+              \ {'lnum' : 4, 'priority' : 20})
+  let s = sign_getplaced('Xsign', {'group' : '*'})
+  call assert_equal(se, s[0].signs)
+  " Place the first sign again with the same priority
+  call sign_place(1, '', 'sign1', 'Xsign',
+              \ {'lnum' : 4, 'priority' : 20})
+  let s = sign_getplaced('Xsign', {'group' : '*'})
+  call assert_equal(se, s[0].signs)
+  " Place the middle sign again with the same priority
+  call sign_place(3, '', 'sign3', 'Xsign',
+              \ {'lnum' : 4, 'priority' : 20})
+  let s = sign_getplaced('Xsign', {'group' : '*'})
+  call assert_equal(se, s[0].signs)
 
   call sign_unplace('*')
```

zeertzjq commented on test/old/testdir/test_signs.vim:
```diff
@@ -1509,42 +1509,46 @@ func Test_sign_priority()
               \ 'priority' : 20}],
               \ s[0].signs)
+
+  " Nvim: signs are always sorted lnum->priority->sign_id->last_modified
```
Hmm, if this order is followed, then `test_termdebug.vim` should need the following change:
```diff
diff --git a/test/old/testdir/test_termdebug.vim b/test/old/testdir/test_termdebug.vim
index 98a4bd321..7fa162c3c 100644
--- a/test/old/testdir/test_termdebug.vim
+++ b/test/old/testdir/test_termdebug.vim
@@ -69,8 +69,7 @@ func Test_termdebug_basic()
         \  'group': 'TermDebug'},
         \ {'lnum': 9, 'id': 1014, 'name': 'debugBreakpoint1.0',
         \  'priority': 110, 'group': 'TermDebug'}],
-        "\ sign_getplaced('', #{group: 'TermDebug'})[0].signs)})
-        \ sign_getplaced('', #{group: 'TermDebug'})[0].signs->reverse())})
+        \ sign_getplaced('', #{group: 'TermDebug'})[0].signs)})
   Finish
   call Nterm_wait(gdb_buf)
   redraw!
```

But currently it doesn't.

luukvbaal replied:
Not familiar with this test and I have no idea how it bypasses the qsort order. Why was it reversed in the first place? I can't reproduce it either with:
```vim
call setline(1, ['a', 'b', 'c']->repeat(3))
call sign_define([{'name':'debugPC', 'text':'>>'}, {'name':'debugBreakpoint1.0', 'text':'xx'}])
call sign_place(12, 'TermDebug', 'debugPC', bufnr(''), {'lnum':9, 'priority':110})
call sign_place(1014, 'TermDebug', 'debugBreakpoint1.0', bufnr(''), {'lnum':9, 'priority':110})
echo sign_getplaced(bufnr(''), {'group':'*'})
```

luukvbaal replied:
Even if I run

    gdb --args <build/bin/nvim> -u unix.vim -U NONE -i NONE --noplugin --headless -u NONE --cmd 'set shortmess-=F' -S runtest.vim test_termdebug.vim

and break in `sign_list_append_info()` I see the expected sign order so I have no idea how this test passes with the incorrect order currently.

luukvbaal replied:
> Hmm, if this order is followed, then test_termdebug.vim should need the following change:

Wait no isn't this just false? The expected results lists them in reverse order so reversing `sign_getplaced()` will make them equal...

zeertzjq replied:
Ah, it's descending order? Hard to notice.

luukvbaal replied:
Yes, ascending line number, descending priority and ID. That doesn't change in this PR. Only thing that changes is that modifying an already placed sign no longer gives it a higher priority in the sign list (see https://github.com/neovim/neovim/pull/25724#discussion_r1368366026).

zeertzjq commented:
Also needs a rebase to remove the unrelated doc change.

zeertzjq commented on runtime/doc/sign.txt:
```diff
@@ -177,11 +175,8 @@ See |sign_place()| for the equivalent Vim script function.
-		each file (if it's used several times unplacing will also have
-		to be done several times and making changes may not work as
-		expected).
+		further manipulation.  {id} must be a number. Placing the same
+		{id} multiple times will move the sign.
```
```suggestion
		further manipulation.  {id} must be a number.  Placing the
		same {id} multiple times will move the sign.
```

zeertzjq commented on src/nvim/sign.c:
```diff
@@ -77,736 +59,245 @@ static char *cmds[] = {
-
-/// Add the sign into the signlist. Find the right spot to do it though.
+/// @param sp  sign properties
+static void buf_addsign(buf_T *buf, int id, char *group, int prio, linenr_T lnum, sign_T *sp)
```
The pointer case below looks a bit risky. I think it's better to cast at caller instead.
```suggestion
static void buf_addsign(buf_T *buf, uint32_t id, char *group, int prio, linenr_T lnum,
                        const sign_T *sp)
```

bfredl commented on src/nvim/sign.c:
```diff
@@ -1758,34 +993,26 @@ static char *get_nth_sign_group_name(int idx)
-        signgroup_T *const group = HI2SG(hi);
-        return group->sg_name;
-      }
+  for (size_t i = 0; i < kv_size(sign_ns); i++) {
```
this can just be a check that `idx` is in bounds `idx < kv_size(sign_ns)` and then use `kv_A(sign_ns, idx)`

luukvbaal commented on src/nvim/decoration.h:
```diff
@@ -60,7 +60,9 @@ struct Decoration {
   int col;  // fixed col value, like win_col
   int virt_text_width;  // width of virt_text
   char *sign_text;
+  char *sign_name;
```
Can we add this to `ExtmarkInfo`? It's helpful for `'statuscolumn'` consumers to know the name attached to a sign mark. So I think `nvim_buf_get_extmarks()` should return the sign name for legacy signs but perhaps we even want to make this a public extmark property?

bfredl replied:
`ExtmarkInfo` is going a way anyway. for now can just access it via the `Decoration *` field, no?

luukvbaal replied:
Yeah adding to `ExtmarkInfo` implies implies adding it to to the `nvim_buf_get_exmarks()` info array, which is what I meant. I can skip `ExtmarkInfo` sure.

bfredl commented on src/nvim/sign.c:
```diff
@@ -77,736 +58,229 @@ static char *cmds[] = {
-  if (sign != NULL) {
-    return sign->se_id;
+  if (kv_size(signs)) {
+    qsort((void *)&kv_A(signs, 0), kv_size(signs), sizeof(MTKey), sign_cmp);
+    extmark_del_id(buf, kv_A(signs, 0).ns, kv_A(signs, 0).id);
+    kv_destroy(signs);
```
What is going on here? The docstring sounds like deleting all signs which matches some specification. If this is intentional, maybe write a small comment why the first item is picked and deleted?

luukvbaal replied:
Yeah intentional. When `atlnum` is specified, only a single (sorted) sign is deleted. There is a comment some lines above:

`  // Store and sort when removing a single sign at a specific line number.`

I can add another here.

bfredl reviewed (approved):
This is great work, and fits very well with the ongoing story of getting the wild growing decoration abstractions (old and new) under control.

As the diffstat indicates this is significant clean-up by itself, and will also enable more simplifications down the way. After getting this merged, my plan is to rebase #25826 onto this and do some further related cleanup.

mckellygit commented:
Hi,
Thanks for awesome nvim.  After this commit I am getting an error in the vim-gitgutter plugin when starting nvim -
```
Error detected while processing /home/user/.vim/bundle/vim-gitgutter/plugin/gitgutter.vim[109]..function gitgutter#highlight#define_signs:
line    1:
E475: Invalid argument:
```
I will debug and try to determine what should change in the plugin, but just wanted to let you know.  Thx.
